chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,263 @@
+
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+type Messages []byte
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships    []map[t.NodeID]t.NodeAddress
+	membershipLock sync.Mutex
+
+	// Channel to send messages to Eudico.
+	ChainNotify chan []Messages
+
+	MirManager *Manager
+
+	reconfigurationVotes map[string]uint64
+}
+
+func NewStateManager(initialMembership map[t.NodeID]t.NodeAddress, m *Manager) *StateManager {
+	// Initialize the membership for the first epochs.
+	// We use configOffset+2 memberships to account for:
+	// - The first epoch (epoch 0)
+	// - The configOffset epochs that already have a fixed membership (epochs 1 to configOffset)
+	// - The membership of the following epoch (configOffset+1) initialized with the same membership,
+	//   but potentially replaced during the first epoch (epoch 0) through special configuration requests.
+	memberships := make([]map[t.NodeID]t.NodeAddress, ConfigOffset+2)
+	for i := 0; i < ConfigOffset+2; i++ {
+		memberships[i] = initialMembership
+	}
+	sm := StateManager{
+		ChainNotify:          make(chan []Messages),
+		MirManager:           m,
+		memberships:          memberships,
+		currentEpoch:         0,
+		reconfigurationVotes: make(map[string]uint64),
+	}
+	return &sm
+}
+
+func (sm *StateManager) ApplyEvents(eventsIn *events.EventList) (*events.EventList, error) {
+	return modules.ApplyEventsSequentially(eventsIn, sm.ApplyEvent)
+}
+
+func (sm *StateManager) ApplyEvent(event *eventpb.Event) (*events.EventList, error) {
+	switch e := event.Type.(type) {
+	case *eventpb.Event_Init:
+		return events.EmptyList(), nil
+	case *eventpb.Event_Deliver:
+		return sm.applyBatch(e.Deliver.Batch)
+	case *eventpb.Event_NewEpoch:
+		return sm.applyNewEpoch(e.NewEpoch)
+	case *eventpb.Event_AppSnapshotRequest:
+		return sm.applySnapshotRequest(e.AppSnapshotRequest)
+	case *eventpb.Event_AppRestoreState:
+		return sm.applyRestoreState(e.AppRestoreState.Snapshot)
+	default:
+		return nil, fmt.Errorf("unexpected type of App event: %T", event.Type)
+	}
+}
+
+// applyBatch applies a batch of requests to the state of the application.
+func (sm *StateManager) applyBatch(in *requestpb.Batch) (*events.EventList, error) {
+	var msgs []Messages
+
+	for _, req := range in.Requests {
+		switch req.Req.Type {
+		case TransportType:
+			msgs = append(msgs, req.Req.Data)
+		case ReconfigurationType:
+			err := sm.applyConfigMsg(req.Req)
+			if err != nil {
+				return events.EmptyList(), err
+			}
+		}
+	}
+
+	// Send messages to the Eudico node.
+	sm.ChainNotify <- msgs
```

Comment:
Maybe you can save yourself the trouble of sending the messages through a channel and submit them to Eudico directly from here (the Eudico API seems to be accessible throught the MirManager here anyway).
It might simplify the code.

Reply:
Yes, it is possible until the state manager contains access to the Mir manager and the Mir manager has access to the Eudico node API.

But:

- it would extend the logic of Mine function over 2 files
- it is not a function of the state manager (within this design) to create concrete blocks (Filecoin blocks) as we discussed
- it will differ from the style of all other Mine functions

So I would prefer to use the current approach at least for now.

Reply:
Fair enough.